File: backend/ddang/src/main/java/com/ddang/ddang/auction/presentation/dto/AuctionResponse.java
package com.ddang.ddang.auction.presentation.dto;

import com.ddang.ddang.auction.application.dto.ReadAuctionDto;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.time.LocalDateTime;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record AuctionResponse(
        Long id,

        String images,

        String title,

        CategoryResponse category,

        String description,

        int startBidPrice,

        Integer lastBidPrice,

        String status,

        int bidUnit,

        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss", timezone = "Asia/Seoul")
        LocalDateTime registerTime,

        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss", timezone = "Asia/Seoul")
        LocalDateTime closingTime,

        DirectRegionResponse directRegions,

        int auctioneerCount,

        boolean deleted
) {

    public static AuctionResponse from(final ReadAuctionDto dto) {
        // TODO 2차 데모데이 이후 리펙터링 예정
        return new AuctionResponse(
                dto.id(),
                dto.image(),
                dto.title(),
                new CategoryResponse(dto.mainCategory(), dto.subCategory()),
                dto.description(),
                dto.startBidPrice(),
                dto.lastBidPrice(),
                "unbidden",
                dto.bidUnit(),
                dto.registerTime(),
                dto.closingTime(),
                new DirectRegionResponse(dto.firstRegion(), dto.secondRegion(), dto.thirdRegion()),
                0,
                dto.deleted()
        );
    }
}

File: android/data/src/main/java/com/ddangddangddang/data/model/response/AuctionResponse.kt
package com.ddangddangddang.data.model.response

import kotlinx.serialization.Serializable

@Serializable
data class AuctionResponse(
    val id: Long,
    val images: List<String>,
    val title: String,
    val category: CategoryResponse,
    val description: String,
    val startPrice: Int,
    val lastBidPrice: Int? = null,
    val status: String,
    val bidUnit: Int,
    val registerTime: String,
    val closingTime: String,
    val directRegions: List<DirectRegionResponse>,
    val auctioneerCount: Int,
    val hasLastBidder: Boolean? = false, // 아직 api 수정 안돼서 일단 이걸로 설정함.
) {
    fun toPreview(): AuctionPreviewResponse = AuctionPreviewResponse(
        id,
        title,
        images.firstOrNull() ?: "",
        lastBidPrice ?: startPrice,
        status,
        auctioneerCount,
    )
}
